### ISSUE TITLE
OCI Integration Improvements

### ISSUE THREAD

**Issue by andresmgot** [kubeapps/kubeapps#2425]
### Description:

After #2232, we noticed several points of improvement for the OCI integration. These have a lower priority, that's why this is a separated issue.

Performance improvements:
 - [x] Avoid checking tags before validating the repository checksum. Each tag requires an API call that we can omit if the repository doesn't include any modifications.
 - [x] Introduce chart versions in the database one by one. At the moment, all the versions are fetched, their files extracted and then stored in the database. This means that for a large number of repositories/tags, all of them needs to be fetched before the user is able to use the repository.
   ~~- [ ] Related to the above, since all the version files are stored in memory, this causes a big memory consumption.~~

UX improvements:
 - [x] When a user is introducing some credentials to access the registry, it'd be helpful to allow using the same credentials for pulling images with one click (without the need of introducing the credentials twice).

**Comment by absoludity** [comment]
Regarding the same credential: if we
* Move the repository type radio field up so it's before the repository authorization section, and
* Switch the repository authorization to enable using (or creating) a secret explicitly (like the later image pull secret) so that the user is aware they're referencing a secret rather than just specifying credentials (we'd talked about doing this already when adding the image pulls secrets),

then we can always create a docker-registry type credential in that case (and still pull out the info we need when using oras or similar). That way, when the user gets to the imagepull secret option below they can just choose a simple "[ ] use the same secret" or similar.

**Comment by andresmgot** [comment]
Taking a look to this again. Regarding the UX for the last, you are talking about something like:

![Screenshot from 2021-03-25 17-10-33](https://user-images.githubusercontent.com/4025665/112505651-4572f900-8d8d-11eb-808b-ea78f5c778a9.png)

The other option would be to add a button when selecting the username and password:

![Screenshot from 2021-03-25 17-19-01](https://user-images.githubusercontent.com/4025665/112506624-3d678900-8d8e-11eb-9c88-f760afbd191c.png)

I think the UX of the second is a bit more friendly? (Regardless of how we store the info as a secret). 

For the first option we can also remove the fields for the server and email in the form to make it a bit more simple (so it's only secret name - username - password).

WDYT?

**Comment by absoludity** [comment]
The key point was that if you select an OCI repo (both your screenshots are helm repos?) for Harbor, you have to enter a username/password to access the private repo and then create a credential for the docker pull secret.

So the key question I have is, is it possible to use a single secret for both? (ie. a secret with a `.dockerconfig`, and our code can adapt to check for both data types when a secret has been specified for accessing an app repo.

If that is possible, then could we do this transparently (but still deterministically) by simply providing a checkbox "[ ] Use this secret for pulling images also" or similar?

**Comment by andresmgot** [comment]
> So the key question I have is, is it possible to use a single secret for both? (ie. a secret with a `.dockerconfig`, and our code can adapt to check for both data types when a secret has been specified for accessing an app repo.
> 
> If that is possible, then could we do this transparently (but still deterministically) by simply providing a checkbox "[ ] Use this secret for pulling images also" or similar?

Yes, we can do that, we need some adaptation in the backend but that's not a problem. The problem is that this form will become more intertwined having inputs affecting other sections. It's also more difficult to understand because this feature will only make sense in none-global namespaces (which are the ones in which you can associate pulling secrets).

I am just trying to figure out the simplest UX possible.

> by simply providing a checkbox "[ ] Use this secret for pulling images also" or similar?

That can be confusing because the secret won't appear in the section for selecting pulling secrets so people may try to create the secret in that section again.

**Comment by absoludity** [comment]
> The problem is that this form will become more intertwined having inputs affecting other sections.

Can't that be easily avoided by switching and putting the checkbox in the section when adding a pull secret, saying "[ ] use the above secret for repository access" or similar, as mentioned in my older comment above?

> It's also more difficult to understand because this feature will only make sense in non-global namespaces (which are the ones in which you can associate pull secrets).

If the checkbox was in the image pull secret section, then it'd only be enabled if that section was enabled, so nothing difficult to understand?

> That can be confusing because the secret won't appear in the section for selecting pull secrets so people may try to create the secret in that section again.

I'd assumed it would be already selected, but either way, I think that's solved by the same change? (Moving the checkbox to the image pull secret section)

You've probably got your head deeper in the UX here, I'm just trying to provide input - hopefully it's helpful :) but I'm ok if you decide otherwise.

**Comment by andresmgot** [comment]
> Can't that be easily avoided by switching and putting the checkbox in the section when adding a pull secret, saying "[ ] use the above secret for repository access" or similar, as mentioned in my older comment above?

The thing is that the section for adding pull secrets is already a checkbox list (you can select several secrets). If we simplify that and only allow to select one secret (so it's a selector):

![Screenshot from 2021-03-26 11-37-42](https://user-images.githubusercontent.com/4025665/112620266-9257dd80-8e28-11eb-924b-445ee222aefc.png)

The downside is that people won't be able to use several secrets (but I don't think no-one is doing that).

WDYT?

> You've probably got your head deeper in the UX here, I'm just trying to provide input - hopefully it's helpful :)

Yes, sure, thanks for the input! I am just struggling with how complex this form is becoming.

**Comment by absoludity** [comment]
> > Can't that be easily avoided by switching and putting the checkbox in the section when adding a pull secret, saying "[ ] use the above secret for repository access" or similar, as mentioned in my older comment above?
> 
> The thing is that the section for adding pull secrets is already a checkbox list (you can select several secrets). If we simplify that and only allow to select one secret (so it's a selector):
> 

It's not obvious to me why the suggested change of providing a checkbox that enables the user to re-use the earlier secret requires limiting the pull secrets to one option? It could remain as is (if we want) and just be something like:
```
[ ] - secretFoo
[ ] - secretBar (same secret as is used for repo access)
[ ] - secretZed

Add a new secret
```

ie. we just ensure we clearly identify the secret which is used for repo access.

> 
> The downside is that people won't be able to use several secrets (but I don't think no-one is doing that).
> 
> WDYT?

Fine either way for me, with a small preference for enabling the same interface that k8s enables (ie. enabling multiple pull secrets).

> 
> > You've probably got your head deeper in the UX here, I'm just trying to provide input - hopefully it's helpful :)
> 
> Yes, sure, thanks for the input! I am just struggling with how complex this form is becoming.

Great. Yep, it should be becoming simplified. AFAICT, enabling the user to select the same secret that they just created can be simpler if we think carefully about the UX.

**Comment by andresmgot** [comment]
> It's not obvious to me why the suggested change of providing a checkbox that enables the user to re-use the earlier secret requires limiting the pull secrets to one option? It could remain as is (if we want) and just be something like:
> 
> ```
> [ ] - secretFoo
> [ ] - secretBar (same secret as is used for repo access)
> [ ] - secretZed
> 
> Add a new secret
> ```
> 
> ie. we just ensure we clearly identify the secret which is used for repo access.
> 
> >

You are suggesting something like:

![Screenshot from 2021-03-29 12-30-22](https://user-images.githubusercontent.com/4025665/112824343-b4e93100-908a-11eb-922d-6e28ef73b4e0.png)

The downsides of that are:

 - The secret must be selected twice (it may be confusing to do so).
 - The form to create secrets is also shown twice (and we cannot remove one of them).

Still prefer the previous option and only allow to select one pullSecret (it's simpler).

**Comment by absoludity** [comment]
> The downsides of that are:
> 
>     * The secret must be selected twice (it may be confusing to do so).

It should be selected twice, shouldn't it? I mean, no matter what we do, the user should be indicating which secret they want to use, even if it means indicating they want to use the same one by implicitly choosing the "Use a docker registry secret"? The UX issue we're trying to avoid is the user having to *enter* the secret twice.

> 
>     * The form to create secrets is also shown twice (and we cannot remove one of them).

You mean the button to "Add new credentials"? Similarly, I'd expect that to be an option in both sections - since I may want to create a credential

> 
> 
> Still prefer the previous option and only allow to select one pullSecret (it's simpler).

Yep, OK - let's go with that then. It is simpler, let's just ensure the wording indicates exactly what's happening (or help text). I mean, if selecting "Use a docker registry secret" implies that you're using the same secret for the repo access and the image pull secrets, that should be clear somehow.

**Comment by andresmgot** [comment]
> It should be selected twice, shouldn't it? I mean, no matter what we do, the user should be indicating which secret they want to use, even if it means indicating they want to use the same one by implicitly choosing the "Use a docker registry secret"? The UX issue we're trying to avoid is the user having to _enter_ the secret twice.
> 

Yes, I agree (selecting the secret twice is better than entering the secret twice), but it's better if you just need to select the secret once.

> Yep, OK - let's go with that then. It is simpler, let's just ensure the wording indicates exactly what's happening (or help text). I mean, if selecting "Use a docker registry secret" implies that you're using the same secret for the repo access and the image pull secrets, that should be clear somehow.

:+1:

**Comment by antgamdia** [comment]
Interesting discussion here. I also go for the simplest solution to avoid a user having to select/write the same thing twice.
However, I'm worried about the overall complexity of this form; it's a bit overwhelming. 
Shouldn't we migrate to a wizard instead? This way we can separate the different concerns, showing the user just the appropriate options.

The problem is that there is no "cds core wizard" yet, just the angular and the clr are available, so perhaps it's better to hold this migration to avoid duplicating efforts.

**Comment by andresmgot** [comment]
Yes, I agree that a wizard would improve the UX here but the impact in dev time is high.

**Comment by absoludity** [comment]
I agree on simplifying the form (+1 to wait for a cds core wizard), a wizard still implies order and atm the thing we're struggling with is that the most obvious order for users is (after choosing whether they want to add an OCI repo or a Helm repo):

1. Set the URI for the repo (and applications for an OCI repo)
2. Select or create a credential to access the repo itself (whether it's an index.yaml or an OCI repo)
3. Select or create a credential to pull images
4. Other advance things like pod spec templates.

It's the interdependence of 2 and 3 which is tricky here, since for some scenarios (Harbor) the same secret can be used for both, while in other scenarios people need a secret only for accessing the repos. Specifically, if people choose a secret for (2) we can't necessarily re-use it for (3) since it may not be a dockerconfigjson secret. I'm not so keen on the current proposal, but don't see a better one, with or without a wizard, to avoid entering the same secret data twice (and hence creating two separate secrets with the same data).

The only way to solve it that I can see would be to reverse (2) and (3) so that the user first chooses whether they need an image pull secrets for the repo, selecting and/or creating those. That way, the next step, choosing whether they need a credential to access the repo itself can simply be selecting the secret from the previous step (I personally think it's fine to select the same secret - it's an intentional decision we should not hide imo) or choosing/creating a different one.

**Comment by andresmgot** [comment]
> The only way to solve it that I can see would be to reverse (2) and (3)

The problem is that (2) is more "important" than (3). Also there is the fact that (3) won't exist for the global namespace.

IMO, we are complicating the form a bit because of the fact that we want to re-use the registry credentials both for accessing the charts and the images.

I think that the simplest solution would be to just have a button to generate Docker credentials from the registry info (and automatically select it in the checkbox list):

![Screenshot from 2021-03-25 17-19-01](https://user-images.githubusercontent.com/4025665/112506624-3d678900-8d8e-11eb-9c88-f760afbd191c.png)

But not having an strong opinion because it's true that with that we would be having the same secret data in two different secrets.

**Comment by absoludity** [comment]
> > The only way to solve it that I can see would be to reverse (2) and (3)
> 
> The problem is that (2) is more "important" than (3).

Not sure it's more important, but it does feel more natural too me too. I'm just not sure if that's because we're used to it being the only option for so long (before we added image pull secret support). If I think as a user trying to add a private repo, it matters less.

> Also there is the fact that (3) won't exist for the global namespace.

Not sure what that would change? The user just doesn't create or set one (and using the same secret is not relevant).

> 
> IMO, we are complicating the form a bit because of the fact that we want to re-use the registry credentials for accessing both the charts and the docker images.

Yes - I agree here: Harbor has bot accounts that have access to both access the helm repo and pull images, and OCI repos work like this generally (I assume?). We're trying to ensure people can add a single credential, even though it's used for two different things for a helm repo. The question is whether we can simplify the form while doing that, or whether it's necessarily more complicated (in which case, perhaps not worth it).

Even for the Harbor OCI repository, I think from memory it's still possible to define one bot account that has access to the OCI repo, and another that has access to pull images (though normally you'd just use one). Otherwise I was going to say that one compromise might be to just add this simplification for OCI repositories (ie. the future) and leave the helm repo as is.

> 
> I think that the simplest solution would be to just have a button to generate Docker credentials from the registry info (and automatically select it in the checkbox list):
> 
> ![Screenshot from 2021-03-25 17-19-01](https://user-images.githubusercontent.com/4025665/112506624-3d678900-8d8e-11eb-9c88-f760afbd191c.png)

I think it's simple to look at, but not necessarily simple to use. For example, I don't think it can be a button, since that implies the docker registry credentials are created when you click, but the auth secret for repo auth is only created when you submit the form, so cancelling the form leaves an inconsistent state. It could instead be a checkbox "[ ] Use the same credential as an image pull secret" which would disable the image pull secret section? That would still work when editing the form too? Why don't you go ahead with your idea (just consider what I mentioned about the checkbox vs button and see what you think).

> 
> But not having an strong opinion because it's true that with that we would be having the same secret data in two different secrets (which btw, is not something wrong because the format and the goal would be different).

Yeah, I agree it's not an issue in and of itself (just a bit icky security-wise).

**Comment by andresmgot** [comment]
Crossing out the 2.1 item since the memory consumption is not that big (at least from the tests I've run). This is the max I got:

![Screenshot from 2021-04-14 16-09-30](https://user-images.githubusercontent.com/4025665/114733613-ca758080-9d43-11eb-8112-380b62aa6b3e.png)

I am going to close this issue but feel free to re-open it if you disagree.
